bin/zbm-kcl

```diff
@@ -206,56 +358,116 @@ done
 
 shift $((OPTIND-1))
 
-zbm_load_lib "zfsbootmenu-kcl.sh"
-zbm_load_lib "echo-log-lib.sh"
-
-bootenv="${1}"
-if [ -z "${bootenv}" ]; then
-  if ! bootenv="$(getbootenv)"; then
+input="${1}"
+if [ -z "${input}" ]; then
+  if ! input="$(getbootenv)"; then
     zerror "root does not appear to be ZFS" >&2
     exit 1
   fi
 fi
 
-if ! zfs list -o name -H "${bootenv}" >/dev/null 2>&1; then
-  zerror "boot environment ${bootenv} does not exist"
-  exit 1
-fi
-
-if [ "${editkcl}" = "yes" ]; then
-  edit "${bootenv}"
-  exit
+if [ -z "${output}" ]; then
+  output="${input}"
 fi
 
+modify=
 if (( ${#removes[@]} != 0 || ${#appends[@]} != 0 )); then
-  havemods="yes"
+  modify="yes"
+elif [ "${editkcl}" = "yes" ]; then
+  modify="yes"
+elif [ "${delkcl}" = "yes" ]; then
+  modify="yes"
 fi
 
-if [ "${delkcl}" ]; then
-  if [ "${havemods}" = "yes" ]; then
-    zerror "-d is incompatible with -a or -r"
+# Make sure to clean up temporary files on exit
+unset TEMP_EFI PRESERVE_EFI TEMP_KCL
+trap cleanup EXIT INT TERM
+
+efi_mode=
+kcl_reader=( "raw_kcl_zfs" "${input}" )
+kcl_writer=( "save_kcl_zfs" "${output}" )
+
+if ! zfs list -o name -H "${input}" >/dev/null 2>&1; then
+  # When input is not a ZFS filesystem, treat it as an EFI executable
+  efi_mode=yes
+
+  if ! TEMP_EFI="$(mktemp)"; then
+    zerror "failed to store EFI executable from stdin"
+    exit 1
+  fi
+
+  if [ "${input}" = "-" ]; then
+    if [ "${editkcl}" = "yes" ]; then
+      zerror "cannot edit KCL in streaming mode"
```

Comment:
What's the impetus behind not supporting editing, on top of appending/removing, in streaming mode? The pieces are all there.

Reply:
If you're streaming edits to a KCL in `std{in,out}`, you're going to be piping things to `zbm-kcl`. There's no clean way to allow an editor to interact with the terminal in that case.